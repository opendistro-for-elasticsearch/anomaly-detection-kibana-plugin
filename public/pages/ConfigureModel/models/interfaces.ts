/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import { AggregationOption } from '../../../models/types';

// Formik values used when creating/editing the model configuration
export interface ModelConfigurationFormikValues {
  featureList: FeaturesFormikValues[];
  categoryFieldEnabled: boolean;
  categoryField: string[];
  shingleSize: number;
}

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string | undefined;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}
