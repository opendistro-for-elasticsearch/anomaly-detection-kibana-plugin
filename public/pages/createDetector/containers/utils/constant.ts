/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { ADFormikValues } from '../models/interfaces';
import { OPERATORS_MAP } from '../../components/DataFilters/utils/constant';
import { UIFilter, FILTER_TYPES } from '../../../../models/interfaces';

export const INITIAL_VALUES: ADFormikValues = {
  detectorName: '',
  timeField: '',
  detectorDescription: '',
  filters: [],
  index: [],
  filterType: FILTER_TYPES.SIMPLE,
  filterQuery: JSON.stringify({ query: { bool: { filter: [] } } }, null, 4),
  detectionInterval: 10,
  windowDelay: 0,
};

export const EMPTY_UI_FILTER: UIFilter = {
  fieldInfo: [],
  operator: OPERATORS_MAP.IS,
  fieldValue: '',
};
