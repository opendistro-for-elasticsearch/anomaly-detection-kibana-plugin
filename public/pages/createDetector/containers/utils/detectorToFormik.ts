/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { cloneDeep, isEmpty, get } from 'lodash';
import { Detector, FILTER_TYPES } from '../../../../models/interfaces';
import { ADFormikValues } from '../models/interfaces';
import { INITIAL_VALUES } from './constant';

export function detectorToFormik(ad: Detector): ADFormikValues {
  const initialValues = cloneDeep(INITIAL_VALUES);
  if (isEmpty(ad)) return initialValues;
  //If detector id updated / added using API, convert all filters to Query boxes as we don't have any meta data.
  const filterType =
    get(ad, 'uiMetadata.filterType', undefined) || FILTER_TYPES.CUSTOM;
  let filterQuery = JSON.stringify(
    get(ad, 'filterQuery', { match_all: {} }),
    null,
    4
  );
  return {
    ...initialValues,
    detectorName: ad.name,
    detectorDescription: ad.description,
    filters: get(ad, 'uiMetadata.filters', []),
    filterType,
    filterQuery,
    index: [{ label: ad.indices[0] }], // Currently we support only one index
    timeField: ad.timeField,
    detectionInterval: get(ad, 'detectionInterval.period.interval', 10),
    windowDelay: get(ad, 'windowDelay.period.interval', 0),
  };
}
