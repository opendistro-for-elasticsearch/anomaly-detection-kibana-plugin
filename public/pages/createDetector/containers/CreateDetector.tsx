/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiFlexGroup,
  EuiFlexItem,
  EuiPage,
  EuiPageBody,
  EuiPageHeader,
  EuiPageHeaderSection,
  EuiSpacer,
  EuiTitle,
} from '@elastic/eui';
import { Formik } from 'formik';
import { get, isEmpty } from 'lodash';
import React, { Fragment, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Dispatch } from 'redux';
//@ts-ignore
import chrome from 'ui/chrome';
//@ts-ignore
import { toastNotifications } from 'ui/notify';
import { APIAction } from '../../../redux/middleware/types';
import {
  createDetector,
  searchDetector,
  updateDetector,
} from '../../../redux/reducers/ad';
import { BREADCRUMBS } from '../../../utils/constants';
import { getErrorMessage, validateName } from '../../../utils/utils';
import { DetectorInfo } from '../components/DetectorInfo';
import { useFetchDetectorInfo } from '../hooks/useFetchDetectorInfo';
import { DataSource } from './DataSource/index';
import { ADFormikValues } from './models/interfaces';
import { detectorToFormik } from './utils/detectorToFormik';
import { formikToDetector } from './utils/formikToDetector';
import { Detector } from '../../../models/interfaces';
import { Settings } from '../components/Settings/Settings';
import { useHideSideNavBar } from '../../main/hooks/useHideSideNavBar';

interface CreateRouterProps {
  detectorId?: string;
}

interface CreateADProps extends RouteComponentProps<CreateRouterProps> {
  isEdit: boolean;
}

export function CreateDetector(props: CreateADProps) {
  useHideSideNavBar(true, false);
  const dispatch = useDispatch<Dispatch<APIAction>>();
  const detectorId: string = get(props, 'match.params.detectorId', '');
  //In case user is refreshing Edit detector page, we'll lose existing detector state
  //This will ensure to fetch the detector based on id from URL
  const { detector, hasError } = useFetchDetectorInfo(detectorId);

  //Set breadcrumbs based on Create / Update
  useEffect(() => {
    const createOrEditBreadcrumb = props.isEdit
      ? BREADCRUMBS.EDIT_DETECTOR
      : BREADCRUMBS.CREATE_DETECTOR;
    let breadCrumbs = [
      BREADCRUMBS.ANOMALY_DETECTOR,
      BREADCRUMBS.DETECTORS,
      createOrEditBreadcrumb,
    ];
    if (detector && detector.name) {
      breadCrumbs.splice(2, 0, {
        text: detector.name,
        //@ts-ignore
        href: `#/detectors/${detectorId}`,
      });
    }
    chrome.breadcrumbs.set(breadCrumbs);
  });
  // If no detector found with ID, redirect it to list
  useEffect(() => {
    if (props.isEdit && hasError) {
      toastNotifications.addDanger('Unable to find detector for edit');
      props.history.push(`/detectors`);
    }
  }, [props.isEdit]);

  const handleUpdate = async (detectorToBeUpdated: Detector) => {
    try {
      await dispatch(updateDetector(detectorId, detectorToBeUpdated));
      toastNotifications.addSuccess(
        `Detector updated: ${detectorToBeUpdated.name}`
      );
      props.history.push(`/detectors/${detectorId}/configurations/`);
    } catch (err) {
      toastNotifications.addDanger(
        getErrorMessage(err, 'There was a problem updating detector')
      );
    }
  };
  const handleCreate = async (detectorToBeCreated: Detector) => {
    try {
      const detectorResp = await dispatch(createDetector(detectorToBeCreated));
      toastNotifications.addSuccess(
        `Detector created: ${detectorResp.data.response.name}`
      );
      props.history.push(
        `/detectors/${detectorResp.data.response.id}/configurations/`
      );
    } catch (err) {
      toastNotifications.addDanger(
        getErrorMessage(err, 'There was a problem creating detector')
      );
    }
  };

  const handleSubmit = async (values: ADFormikValues, formikBag: any) => {
    const apiRequest = formikToDetector(values, detector);
    try {
      if (props.isEdit) {
        await handleUpdate(apiRequest);
      } else {
        await handleCreate(apiRequest);
      }
      formikBag.setSubmitting(false);
    } catch (e) {
      formikBag.setSubmitting(false);
    }
  };

  const handleCancelClick = () => {
    props.history.push('/detectors');
  };

  const handleValidateName = async (detectorName: string) => {
    const {
      isEdit,
      match: {
        params: { detectorId },
      },
    } = props;
    if (isEmpty(detectorName)) {
      throw 'Detector name can not be empty';
    } else {
      const error = validateName(detectorName);
      if (error) {
        throw error;
      }
      //TODO::Avoid making call if value is same
      const resp = await dispatch(
        searchDetector({ query: { term: { 'name.keyword': detectorName } } })
      );
      const totalDetectors = resp.data.response.totalDetectors;
      if (totalDetectors === 0) {
        return undefined;
      }
      //If more than one detectors found, duplicate exists.
      if (!isEdit && totalDetectors > 0) {
        throw 'Duplicate detector name';
      }
      // if it is in edit mode
      if (
        isEdit &&
        (totalDetectors > 1 ||
          get(resp, 'data.response.detectors.0.id', '') !== detectorId)
      ) {
        throw 'Duplicate detector name';
      }
    }
  };

  return (
    <EuiPage>
      <EuiPageBody>
        <EuiPageHeader>
          <EuiPageHeaderSection>
            <EuiTitle size="l">
              <h1>{props.isEdit ? 'Edit detector' : 'Create detector'} </h1>
            </EuiTitle>
          </EuiPageHeaderSection>
        </EuiPageHeader>
        <Formik
          enableReinitialize={true}
          initialValues={detectorToFormik(detector)}
          onSubmit={handleSubmit}
        >
          {(formikProps) => (
            <Fragment>
              <DetectorInfo onValidateDetectorName={handleValidateName} />
              <EuiSpacer />
              <DataSource formikProps={formikProps} />
              <EuiSpacer />
              <Settings />
              <EuiSpacer />
              <EuiFlexGroup alignItems="center" justifyContent="flexEnd">
                <EuiFlexItem grow={false}>
                  <EuiButtonEmpty onClick={handleCancelClick}>
                    Cancel
                  </EuiButtonEmpty>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiButton
                    fill
                    type="submit"
                    isLoading={formikProps.isSubmitting}
                    //@ts-ignore
                    onClick={formikProps.handleSubmit}
                  >
                    {props.isEdit ? 'Save changes' : 'Create'}
                  </EuiButton>
                </EuiFlexItem>
              </EuiFlexGroup>
            </Fragment>
          )}
        </Formik>
      </EuiPageBody>
    </EuiPage>
  );
}
