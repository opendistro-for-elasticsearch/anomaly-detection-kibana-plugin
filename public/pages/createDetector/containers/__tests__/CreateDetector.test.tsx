/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import {
  HashRouter as Router,
  RouteComponentProps,
  Route,
  Switch,
} from 'react-router-dom';
import { render, fireEvent, wait } from '@testing-library/react';
// @ts-ignore
import { toastNotifications } from 'ui/notify';
import { CreateDetector } from '../CreateDetector';
import { getRandomDetector } from '../../../../redux/reducers/__tests__/utils';
import configureStore from '../../../../redux/configureStore';
import { httpClientMock } from '../../../../../test/mocks';
import userEvent from '@testing-library/user-event';
import { UNITS } from '../../../../models/interfaces';

const renderWithRouter = (isEdit: boolean = false) => ({
  ...render(
    <Provider store={configureStore(httpClientMock)}>
      <Router>
        <Switch>
          <Route
            render={(props: RouteComponentProps) => (
              <CreateDetector isEdit={isEdit} {...props} />
            )}
          />
        </Switch>
      </Router>
    </Provider>
  ),
});

describe('<CreateDetector /> spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('create detector', () => {
    test('renders the component', () => {
      httpClientMock.get = jest.fn().mockResolvedValue({
        data: {
          ok: true,
          response: {
            indices: [
              { index: 'hello', health: 'green' },
              { index: 'world', health: 'yellow' },
            ],
          },
        },
      });
      const { container } = renderWithRouter();
      expect(container.firstChild).toMatchSnapshot();
    });

    test('validate all required field', async () => {
      httpClientMock.get = jest.fn().mockResolvedValue({
        data: {
          ok: true,
          response: {
            indices: [
              { index: 'hello', health: 'green' },
              { index: 'world', health: 'yellow' },
            ],
          },
        },
      });
      const { getByText } = renderWithRouter();
      fireEvent.click(getByText('Create'));
      await wait(() => {
        getByText('Required');
        getByText('Must specify an index');
      });
    });

    test('prevent duplicate detector name', async () => {
      const randomDetector = getRandomDetector();
      httpClientMock.post = jest.fn().mockResolvedValue({
        data: {
          ok: true,
          response: { detectors: [randomDetector], totalDetectors: 1 },
        },
      });
      const { getByPlaceholderText, getByText } = renderWithRouter();
      fireEvent.focus(getByPlaceholderText('Enter detector name'));
      userEvent.type(
        getByPlaceholderText('Enter detector name'),
        randomDetector.name
      );
      fireEvent.blur(getByPlaceholderText('Enter detector name'));
      await wait(() => {
        getByText('Duplicate detector name');
      });
    });
    //There are some issues / bug with Combobox and unable to select item
    it.skip('should create detector with all valid values', async () => {
      const randomDetector = getRandomDetector();
      //Mocked name validation search detector
      httpClientMock.post = jest.fn().mockResolvedValue({
        data: {
          ok: true,
          response: { detectors: [randomDetector], totalDetectors: 1 },
        },
      });
      //Mocked get indices
      httpClientMock.get = jest.fn().mockResolvedValue({
        data: {
          ok: true,
          response: {
            indices: [
              { index: 'hello', health: 'green' },
              { index: 'world', health: 'yellow' },
            ],
          },
        },
      });
      const { getByPlaceholderText } = renderWithRouter();
      fireEvent.focus(getByPlaceholderText('sample detector'));
      userEvent.type(
        getByPlaceholderText('sample detector'),
        randomDetector.name
      );
      fireEvent.focus(getByPlaceholderText('Description for detector'));
      userEvent.type(
        getByPlaceholderText('Description for detector'),
        randomDetector.description
      );
    });
  });
});
