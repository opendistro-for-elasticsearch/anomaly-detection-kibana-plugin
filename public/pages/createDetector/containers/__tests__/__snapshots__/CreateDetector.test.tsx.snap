// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateDetector /> spec create detector renders the component 1`] = `
<div
  class="euiPage"
>
  <div
    class="euiPageBody"
  >
    <div
      class="euiPageHeader euiPageHeader--responsive"
    >
      <div
        class="euiPageHeaderSection"
      >
        <h1
          class="euiTitle euiTitle--large"
        >
          Create detector
           
        </h1>
      </div>
    </div>
    <div
      class="euiPanel euiPanel--paddingMedium"
      style="padding: 20px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px;"
      >
        <div
          class="euiFlexItem"
        >
          <h3
            class="euiTitle euiTitle--small content-panel-title"
          >
            Name and description
          </h3>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            />
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            />
          </div>
        </div>
      </div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiFormRow"
          hint="Specify a unique and descriptive name that is easy to recognize. Detector name must contain 1-64 characters.  Valid characters are a-z, A-Z, 0-9, -(hyphen) and _(underscore)"
          id="random_id-row"
          title="Name"
        >
          <div>
            <label
              aria-invalid="false"
              class="euiFormLabel"
              for="random_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Name
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Specify a unique and descriptive name that is easy to recognize. Detector name must contain 1-64 characters.  Valid characters are a-z, A-Z, 0-9, -(hyphen) and _(underscore)
                </p>
              </div>
            </label>
          </div>
          <div
            class="euiFormControlLayout"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                class="euiFieldText"
                id="random_id"
                name="detectorName"
                placeholder="Enter detector name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="euiFormRow"
          formattedtitle="[object Object]"
          hint="Describe the purpose of the detector."
          id="random_id-row"
        >
          <div>
            <label
              aria-invalid="false"
              class="euiFormLabel"
              for="random_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Description 
                  <span
                    class="optional"
                  >
                    - optional
                  </span>
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Describe the purpose of the detector.
                </p>
              </div>
            </label>
          </div>
          <textarea
            class="euiTextArea euiTextArea--resizeVertical"
            id="random_id"
            name="detectorDescription"
            placeholder="Describe the detector"
            rows="3"
          />
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiPanel euiPanel--paddingMedium"
      style="padding: 20px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px;"
      >
        <div
          class="euiFlexItem"
        >
          <h3
            class="euiTitle euiTitle--small content-panel-title"
          >
            Data Source
          </h3>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            />
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            />
          </div>
        </div>
      </div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiFormRow"
          hint="Choose an index or index pattern as the data source."
          id="random_id-row"
          title="Index"
        >
          <div>
            <label
              aria-invalid="false"
              class="euiFormLabel"
              for="random_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Index
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Choose an index or index pattern as the data source.
                </p>
              </div>
            </label>
          </div>
          <div
            aria-describedby="random_id-help"
            aria-expanded="false"
            aria-haspopup="listbox"
            class="euiComboBox"
            role="combobox"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <div
                  class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap"
                  data-test-subj="comboBoxInput"
                  tabindex="-1"
                >
                  <p
                    class="euiComboBoxPlaceholder"
                  >
                    Find indices
                  </p>
                  <div
                    class="euiComboBox__input"
                    style="font-size: 14px; display: inline-block;"
                  >
                    <input
                      data-test-subj="comboBoxSearchInput"
                      id="random_id"
                      role="textbox"
                      style="box-sizing: content-box; width: 2px;"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <button
                    aria-label="Open list of options"
                    class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                    data-test-subj="comboBoxToggleListButton"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                      focusable="false"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFormHelpText euiFormRow__text"
            id="random_id-help"
          >
            You can use a wildcard (*) in your index pattern
          </div>
        </div>
        <div
          class="euiFormRow"
          hint="Choose the time field you want to use for time filter."
          id="random_id-row"
          title="Timestamp field"
        >
          <div>
            <label
              aria-invalid="false"
              class="euiFormLabel"
              for="random_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Timestamp field
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Choose the time field you want to use for time filter.
                </p>
              </div>
            </label>
          </div>
          <div
            class="euiFormControlLayout"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <select
                class="euiSelect"
                id="random_id"
                name="timeField"
                placeholder="Find timestamp"
              >
                <option
                  value=""
                />
              </select>
              <div
                class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
              >
                <span
                  class="euiFormControlLayoutCustomIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                    focusable="false"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFormRow euiFormRow--fullWidth"
          formattedtitle="[object Object]"
          hint="Choose a subset of your data source to focus your data stream and reduce noisy data.,Use the visual editor to create a simple filter, or use the Elasticsearch query DSL to create more advanced filters."
          id="random_id-row"
        >
          <div>
            <label
              aria-invalid="false"
              class="euiFormLabel"
              for="random_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Data fiter
                  <span
                    class="optional"
                  >
                    - optional
                  </span>
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Choose a subset of your data source to focus your data stream and reduce noisy data.
                </p>
                <p
                  class="sublabel"
                >
                  Use the visual editor to create a simple filter, or use the Elasticsearch query DSL to create more advanced filters.
                </p>
              </div>
            </label>
          </div>
          <div
            class="euiFormControlLayout"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <select
                class="euiSelect"
                id="random_id"
                name="filterType"
              >
                <option
                  value="simple_filter"
                >
                  Visual editor
                </option>
                <option
                  value="custom_filter"
                >
                  Custom expression
                </option>
              </select>
              <div
                class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
              >
                <span
                  class="euiFormControlLayoutCustomIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                    focusable="false"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full"
        />
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 10px;"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <button
              class="euiButton euiButton--primary"
              type="button"
            >
              <span
                class="euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Add filter
                </span>
              </span>
            </button>
          </div>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full"
        />
        <div
          class="euiPanel euiPanel--paddingMedium"
          style="background-color: rgb(246, 246, 246);"
        >
          <div
            class="no-data-filter-rectangle"
          >
            <div
              class="euiText euiText--medium"
            >
              <p
                class="no-data-filter-title"
              >
                No data filter
              </p>
            </div>
            <div
              class="euiSpacer euiSpacer--s"
            />
            <div
              class="euiText euiText--medium"
            >
              <p
                class="sublabel-center"
              >
                Use data filter to reduce noisy data
              </p>
            </div>
            <div
              class="euiSpacer euiSpacer--s"
            />
            <button
              class="euiButton euiButton--primary"
              type="button"
            >
              <span
                class="euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Add filter
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiPanel euiPanel--paddingMedium"
      style="padding: 20px;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px;"
      >
        <div
          class="euiFlexItem"
        >
          <h3
            class="euiTitle euiTitle--small content-panel-title"
          >
            Detector operation settings
          </h3>
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            />
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            />
          </div>
        </div>
      </div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
            style="max-width: 70%;"
          >
            <div
              class="euiFormRow euiFormRow--fullWidth"
              hint="Define how often the detector collects data to generate anomalies. The shorter the interval is, the more real time the detector results will be, and the more computing resources the detector will need. The detector interval, together with data ingestion speed, also determines the preparation time of the detector. Long intervals and slow ingestion speed means the detector takes longer time to collect sufficient data to generate anomalies."
              id="random_id-row"
              title="Detector interval"
            >
              <div>
                <label
                  aria-invalid="false"
                  class="euiFormLabel"
                  for="random_id"
                >
                  <div
                    style="line-height: 8px;"
                  >
                    <p>
                      Detector interval
                    </p>
                    <br />
                    <p
                      class="sublabel"
                    >
                      Define how often the detector collects data to generate anomalies. The shorter the interval is, the more real time the detector results will be, and the more computing resources the detector will need. The detector interval, together with data ingestion speed, also determines the preparation time of the detector. Long intervals and slow ingestion speed means the detector takes longer time to collect sufficient data to generate anomalies.
                    </p>
                  </div>
                </label>
              </div>
              <div
                class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                id="random_id"
              >
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  <div
                    class="euiFormControlLayout"
                  >
                    <div
                      class="euiFormControlLayout__childrenWrapper"
                    >
                      <input
                        class="euiFieldNumber"
                        data-test-subj="detectionInterval"
                        id="detectionInterval"
                        name="detectionInterval"
                        placeholder="Detector interval"
                        type="number"
                        value="10"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="euiFlexItem"
                >
                  <div
                    class="euiText euiText--medium"
                  >
                    <p
                      class="minutes"
                    >
                      minutes
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFormRow euiFormRow--fullWidth"
          hint="Specify a window of delay for a detector to fetch data, if you need to account for extra processing time."
          id="random_id-row"
          style="margin-top: 16px;"
          title="Window delay"
        >
          <div>
            <label
              aria-invalid="false"
              class="euiFormLabel"
              for="random_id"
            >
              <div
                style="line-height: 8px;"
              >
                <p>
                  Window delay
                </p>
                <br />
                <p
                  class="sublabel"
                >
                  Specify a window of delay for a detector to fetch data, if you need to account for extra processing time.
                </p>
              </div>
            </label>
          </div>
          <div
            class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
            id="random_id"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldNumber"
                    data-test-subj="windowDelay"
                    id="windowDelay"
                    name="windowDelay"
                    placeholder="Window delay"
                    type="number"
                    value="1"
                  />
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiText euiText--medium"
              >
                <p
                  class="minutes"
                >
                  minutes
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <button
          class="euiButtonEmpty euiButtonEmpty--primary"
          type="button"
        >
          <span
            class="euiButtonEmpty__content"
          >
            <span
              class="euiButtonEmpty__text"
            >
              Cancel
            </span>
          </span>
        </button>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <button
          class="euiButton euiButton--primary euiButton--fill"
          type="submit"
        >
          <span
            class="euiButton__content"
          >
            <span
              class="euiButton__text"
            >
              Create
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
